sdk/go/common/workspace/plugins.go

```diff
@@ -344,6 +348,98 @@ func (source *gitlabSource) URL() string {
 	return fmt.Sprintf("gitlab://%s/%s", source.host, source.project)
 }
 
+type gitSource struct {
+	url  string
+	path string
+
+	cloneAndCheckout func(context.Context, string, plumbing.Hash, string) error
+	cloneOrPull      func(context.Context, string, plumbing.ReferenceName, string, bool) error
+}
+
+func newGitHTTPSSource(url *url.URL) (*gitSource, error) {
+	url.Scheme = "https"
+	u, path, err := gitutil.ParseGitRepoURL(url.String())
+	if err != nil {
+		return nil, err
+	}
+	if path != "" {
+		return nil,
+			// TODO[pulumi/pulumi#18250]: We currently don't support installing plugins from subdirectories.
+			//nolint:lll
+			errors.New("cannot install git plugin from subdirectory.  See https://github.com/pulumi/pulumi/issues/18250")
+	}
+	return &gitSource{
+		url:              u,
+		path:             path,
+		cloneAndCheckout: gitutil.GitCloneAndCheckoutCommit,
+		cloneOrPull:      gitutil.GitCloneOrPull,
+	}, nil
+}
+
+func (source *gitSource) GetLatestVersion(
+	ctx context.Context, _ func(*http.Request) (io.ReadCloser, int64, error),
+) (*semver.Version, error) {
+	// TODO: support this for the users convenience
+	return &semver.Version{},
+		errors.New("GetLatestVersion is not supported for git sources, please specify the version of the plugin")
+}
+
+func (source *gitSource) Download(
+	ctx context.Context, version semver.Version, _ string, _ string,
+	_ func(*http.Request) (io.ReadCloser, int64, error),
+) (io.ReadCloser, int64, error) {
+	tmpdir, err := os.MkdirTemp("", "pulumi-plugin")
+	if err != nil {
+		return nil, -1, err
+	}
+	defer os.RemoveAll(tmpdir)
+	if version.Major == 0 && version.Minor == 0 && version.Patch == 0 && len(version.Pre) > 0 {
```

Comment:
I think this logic appears again later. Perhaps worth factoring out an `isPrereleaseVersion` or something?